Rework editor docks

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/12435
Fixes #106927

There are some differences from the proposal. The new class is called EditorDock and inherits MarginContainer.

I still need to expose it. For now I changed all docks to use the new system.

EDIT:
Done. I made an example plugin that uses the new system:
[dock-test.zip](https://github.com/user-attachments/files/20534966/dock-test.zip)
It can be moved to bottom and supports saving layout.

Also supersedes #106428

KoBeWi committed:
Rework editor docks

lodetrick commented:
This isn't an issue with this PR, but since we're changing how the dock customization works would it be possible to take #106428 into account? In my opinion it is better to have the brittle Texture2D to be the priority rather than the fallback because it would allow for overrides and the more dynamic StringName (it is able to be drawn in dark mode, for example) to be the default.

KoBeWi commented:
Should be ready now. I included #106428 too.
Makes me wonder whether `icon_name` should be renamed to `fallback_icon` maybe?

lodetrick commented:
I still think the icon_name has a use because it is a lot more flexible (it can be used in light or dark mode), maybe it could be `icon` -> `override_icon` and `icon_name` -> `default_icon`.

KeyboardDanni commented:
I mentioned this already in the Godot Contributors Chat, but I'm concerned about every dock being added to the menubar under "Editor Docks". The problem is that this shows every possible context-sensitive dock to the user, even when they wouldn't normally be visible. For example, the "Tilemap" dock usually only appears when you are editing a Tilemap. I'm not exactly sure under what circumstances you'd want to access that dock when you're not editing a Tilemap. As goofy as the Microsoft Office Ribbon is, I feel that this is one aspect they did get right at least. In versions of Microsoft Office before 2007 the UI was crammed with a lot of toolbars, docks, selectable windows, that sort of thing, for features that aren't even active so they're just taking up space.

I could see the need to be able to access a dock that isn't context-relevant if you want to adjust where it is in the layout, but I'm not sure how best to approach that. Some programs temporarily show any stuff that would normally be hidden when you go into the "layout edit mode" but Godot doesn't really have that.

I do like the idea of being able to move docks that would normally be bottom-locked, and put them in other parts of the UI. But also a lot of bottom docks are situational, and them being at the bottom communicates that. So moving them elsewhere doesn't really fit the established paradigm.

KoBeWi commented:
That's unrelated to this PR though. It only changes internals, the editor so far works the same as before.

lodetrick commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -1145,14 +1146,18 @@ VersionControlEditorPlugin::VersionControlEditorPlugin() {
+	version_commit_dock = memnew(EditorDock);
 	version_commit_dock->set_visible(false);
-	version_commit_dock->set_name(TTR("Commit"));
+	version_commit_dock->set_name(TTRC("Commit"));
+	version_commit_dock->set_dock_shortcut(ED_SHORTCUT_AND_COMMAND("docks/open_version_control", TTRC("Open Version Control Dock")));
```
```suggestion
	version_commit_dock->set_name(TTRC("Commit"));
	version_commit_dock-> set_icon_name("VcsBranches");
	version_commit_dock->set_dock_shortcut(ED_SHORTCUT_AND_COMMAND("docks/open_version_control", TTRC("Open Version Control Dock")));
```
This isn't strictly necessary but the version control dock doesn't currently have an icon, so it will probably default to the "window" icon.

lodetrick reviewed (commented):
The main content looked good from what I could see, although there were quite a few object leaks: 
```
WARNING: 5 RIDs of type "CanvasItem" were leaked.
     at: _free_rids (servers/rendering/renderer_canvas_cull.cpp:2678)
ERROR: 6 RID allocations of type 'N10RendererRD14TextureStorage7TextureE' were leaked at exit.
WARNING: 8 RIDs of type "Texture" were leaked.
     at: finalize (servers/rendering/rendering_device.cpp:7164)
ERROR: 73 RID allocations of type 'PN18TextServerAdvanced22ShapedTextDataAdvancedE' were leaked at exit.
ERROR: 18 RID allocations of type 'PN18TextServerAdvanced12FontAdvancedE' were leaked at exit.
ERROR: 1 RID allocations of type 'PN18TextServerAdvanced27FontAdvancedLinkedVariationE' were leaked at exit.
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object/object.cpp:2489)
ERROR: 2 resources still in use at exit (run with --verbose for details).
   at: clear (core/io/resource.cpp:637)
```

It might be an issue with master. I'll check. Yeah it has to do with master. Disregard above.

lodetrick commented on editor/plugins/editor_plugin.cpp:
```diff
@@ -84,24 +85,46 @@ Button *EditorPlugin::add_control_to_bottom_panel(Control *p_control, const Stri
-	EditorDockManager::get_singleton()->remove_dock(p_control);
+	ERR_FAIL_COND(!legacy_docks.has(p_control));
+
+	EditorDockManager::get_singleton()->remove_dock(legacy_docks[p_control]);
+	legacy_docks.erase(p_control);
```
Also would it make sense to make `legacy_docks` static?

lodetrick commented on editor/plugins/editor_plugin.h:
```diff
@@ -140,6 +143,10 @@ class EditorPlugin : public Node {
 	Button *_add_control_to_bottom_panel_compat_88081(Control *p_control, const String &p_title);
 	void _add_control_to_dock_compat_88081(DockSlot p_slot, Control *p_control);
 	static void _bind_compatibility_methods();
+
+	void add_control_to_dock(DockSlot p_slot, Control *p_control, const Ref<Shortcut> &p_shortcut = nullptr);
+	void remove_control_from_docks(Control *p_control);
```
It might be an issue that these functions are protected now, as they were public in the past. Or should they have been protected from the start?

KoBeWi replied:
Seems like the compatibility methods are also protected.
It does not really matter, it only affects C++ modules that call plugins methods from outside the plugin, but there is no reason to do so.

lodetrick replied:
Makes me curious why they were public to begin with

KoBeWi replied:
Protected methods are just rarely used in the engine. There are some common protected methods like `notification()` or `bind_method()` and then everything else is either private or public.

lodetrick commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,95 @@
+		class MyDock extends EditorDock:
+		    func _init():
+		        title = "My Dock"
+		        icon = load("./dock_icon.png")
```
```suggestion
		        dock_icon = load("./dock_icon.png")
```

lodetrick commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,95 @@
+		# Plugin initialization.
+		func _enter_tree():
+		    dock = MyDock.new()
+		    add_dock(dock)
```
```suggestion
		    add_dock(dock, EditorPlugin.DOCK_SLOT_RIGHT_UL)
```

lodetrick commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,95 @@
+		        title = "My Dock"
+		        icon = load("./dock_icon.png")
+
+		        var dock_content = preload("./dock_content.tscn").instantiate()
```
```suggestion
		        var dock_content = preload("res://dock_content.tscn").instantiate()
```

KoBeWi replied:
That one is intended. The scene is supposed to be loaded from addon's directory and relative paths are supported in `preload()`.

lodetrick replied:
Thanks for the clarification, I'm not too familiar with plugins specifically

lodetrick commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,95 @@
+		EditorDock is a [Container] node that can be docked in one of the editor's dock slots. Docks are added by plugins to provide space for controls related to the plugin.
+		You can add a dock by using [method EditorPlugin.add_dock]. The dock can be customized by changing its properties.
+		[codeblock]
+		extends EditorPlugin
```
```suggestion
		@tool
		extends EditorPlugin
```
Should we include the `@tool` or do plugins not need them every time

lodetrick commented on editor/editor_dock.cpp:
```diff
@@ -0,0 +1,133 @@
+}
+
+void EditorDock::set_dock_shortcut(const Ref<Shortcut> &p_shortcut) {
+	shortcut = p_shortcut;
```
Shortcuts are only updated when docks are added, removed, or moved right now, so any change to this while the dock is added will not come into effect immediately.

KoBeWi replied:
I added a note.

lodetrick reviewed (approved):
It looks good now from what I can see. Thanks for your effort!

kitbdev commented on editor/editor_dock_manager.cpp:
```diff
@@ -461,11 +467,11 @@ void EditorDockManager::save_docks_to_config(Ref<ConfigFile> p_layout, const Str
 		window_dump["window_screen_rect"] = DisplayServer::get_singleton()->screen_get_usable_rect(screen);
 
-		String name = dock->get_name();
+		String name = dock->get_display_title();
```
We should not use the display title to save/load data. It's now pretty easy to change the title at any time for a dock, so this value cannot be relied on. Saving and loading needs this to be the same or it will not work.

kitbdev replied:
The previous code looks like it used the control's name and just ignored the title, since they always effectively had a name before.
I thought it was even encouraged now because the `tab_style_changed` signal emits when the title changes so it updates properly. 
And though it's not a dock, the Debugger bottom panel changes it's display to `Debugger (1)` to show the error count dynamically.
We could just have another property to get a name for saving, and if it is empty fall back to the display title.

lodetrick commented:
We should probably also clip the contents of the docks:
<img width="641" alt="Not Clipped" src="https://github.com/user-attachments/assets/ffa4eaa9-5c59-4f94-bd2d-80fdd6abec12" />

KoBeWi commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,96 @@
+		@tool
+		extends EditorPlugin
+
+		# Define the dock class (can be in separate file).
```
EditorDock extends MarginContainer, which does not alter appearance of its children. Your dock scene will appear the same when added under dock.

KoBeWi commented on editor/editor_dock.cpp:
```diff
@@ -0,0 +1,133 @@
+#include "core/input/shortcut.h"
+#include "core/io/config_file.h"
+
+void EditorDock::_bind_methods() {
```
It's the same order as in `.h`.

kitbdev replied:
It's the same order, but most other files are not.
Well it doesn't really matter.

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -273,45 +272,44 @@ void EditorDockManager::_open_dock_in_window(Control *p_dock, bool p_show_window
-	p_dock->call("_set_dock_horizontal", true);
+	p_dock->at_bottom = true;
+	p_dock->previous_at_bottom = false;
+	p_dock->set_horizontal_layout(true);
```
I mimicked what FileSystem dock currently does. This can be changed later without breaking compatibility.

KoBeWi commented on editor/editor_dock.cpp:
```diff
@@ -0,0 +1,133 @@
+	GDVIRTUAL_CALL(_set_horizontal_layout, p_enable);
+}
+
+void EditorDock::save_layout_to_config(Ref<ConfigFile> &p_layout, const String &p_section) const {
+	GDVIRTUAL_CALL(_save_layout_to_config, p_layout, p_section);
+}
+
+void EditorDock::load_layout_from_config(const Ref<ConfigFile> &p_layout, const String &p_section) {
+	GDVIRTUAL_CALL(_load_layout_from_config, p_layout, p_section);
+}
```
Using the provided section ensures that you don't modify any unintended data. I could add a note in the docs.
We already have `_get_window_layout()`, which provides ConfigFile (and it's worse, because you don't get a dedicated section).

kitbdev replied:
It uses the same section as the dock manager, so if you use a value like "dock_1" it will overwrite the existing value and mess up the layout.

KoBeWi replied:
Right, maybe I should make every dock get their dedicated section 🤔

KoBeWi commented on editor/editor_dock_manager.h:
```diff
@@ -144,22 +131,20 @@ class EditorDockManager : public Object {
-	void remove_dock(Control *p_dock);
-
-	void set_dock_tab_icon(Control *p_dock, const Ref<Texture2D> &p_icon);
+	void add_dock(EditorDock *p_dock, DockSlot p_slot = DOCK_SLOT_NONE);
```
But after the dock is added, should changing the property move it?

kitbdev replied:
Probably not, the user should control where docks are after the first time they open.

KoBeWi commented on editor/plugins/editor_plugin.h:
```diff
@@ -148,11 +155,10 @@ class EditorPlugin : public Node {
 	void remove_control_from_bottom_panel(Control *p_control);
 
-	void set_dock_tab_icon(Control *p_control, const Ref<Texture2D> &p_icon);
+	void add_dock(EditorDock *p_dock, DockSlot p_slot);
+	void remove_dock(EditorDock *p_dock);
```
It would be difficult to manage outside EditorPlugin. You need to handle adding and deleting the dock.

KoBeWi commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,96 @@
+			Allows placing the dock in the vertical dock slots on either of the editor.
+			[b]Note:[/b] Currently this flag has no effect, because the bottom panel is not a proper dock slot (yet). The dock can always be vertical.
+		</constant>
+		<constant name="DOCK_LAYOUT_HORIZONTAL" value="2" enum="DockLayout" is_bitfield="true">
+			Allows placing the dock in the editor's bottom panel. Implement [method _set_horizontal_layout] to handle changing layouts.
```
Bottom panel is currently the only place for horizontal docks, but this might change.

Mickeon commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,96 @@
+			The icon for the dock. If specified, it will override [member icon_name].
+		</member>
+		<member name="dock_shortcut" type="Shortcut" setter="set_dock_shortcut" getter="get_dock_shortcut">
+			The shortcut used to open the dock. Setting this is only reliable before the dock is added.
```
I am genuinely not sure what "reliable" means here. I am to assume:
```suggestion
			The shortcut used to open the dock. This property can only be set before this dock is added via [method EditorPlugin.add_dock].
```

KoBeWi replied:
Your new description suggests that you can't set the property after adding. You can, but it won't take effect until you re-add the dock.

Mickeon commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,96 @@
+			The shortcut used to open the dock. Setting this is only reliable before the dock is added.
+		</member>
+		<member name="icon_name" type="StringName" setter="set_icon_name" getter="get_icon_name" default="&amp;&quot;&quot;">
+			The icon name for the dock. It has to use an icon name from [code]EditorIcons[/code] in editor theme. You can find the list of available icons [url=https://godot-editor-icons.github.io/]here[/url].
```
What's `EditorIcons` in this context? "From the `EditorIcons` type"...? "The `EditorIcons` variation"...?

KoBeWi commented:
Updated. From more important changes:
- I added `default_slot` property, which replaces the slot argument of `add_dock()`
- I added `layout_key` property, which is a string representing this dock in the layout file
- Changed `_set_horizontal_layout()` to `_update_layout()`, the argument is now the new layout
  - This will be useful if we decide to add more layouts (maybe Floating or something)

kitbdev reviewed (commented):
Saving and loading do not happen when the dock is added or removed, this can cause data loss since adding a dock will have default values and it will overwrite the existing values, or it can be removed before it is saved.
Is it possible to save or load only 1 dock? We may want to trigger a save before a dock is removed and a load after a dock is added, somehow.
This could be fixed later though, since that is the existing behavior.

KoBeWi commented on doc/classes/FileSystemDock.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<class name="FileSystemDock" inherits="VBoxContainer" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+<class name="FileSystemDock" inherits="EditorDock" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
```
All checks have passed, so no idea. I doubt anyone would use FileSystemDock in a way that it would cause problem.

KoBeWi commented on editor/editor_dock.h:
```diff
@@ -0,0 +1,113 @@
+protected:
+	static void _bind_methods();
+
+	GDVIRTUAL1(_update_layout, int)
```
```
[ 99%] C:\godot_source\editor/editor_dock.h(77): error C2027: wykorzystanie niezdefiniowanego przez użytkownika typu "GetTypeInfo<EditorDock::DockLayout,void>"
C:\godot_source\core/variant/type_info.h(62): note: zobacz deklarację „GetTypeInfo<EditorDock::DockLayout,void>”
C:\godot_source\editor/editor_dock.h(77): error C3861: "get_class_info": identyfikatora nie odnaleziono
C:\godot_source\editor/editor_dock.h(77): error C2065: "METADATA": niezadeklarowany identyfikator
scons: *** [bin\obj\editor\gui\.scu\scu_editor_gui.gen.windows.editor.dev.x86_64.obj] Error 2
C:\godot_source\editor/editor_dock.h(77): error C2027: wykorzystanie niezdefiniowanego przez użytkownika typu "GetTypeInfo<EditorDock::DockLayout,void>"
C:\godot_source\core/variant/type_info.h(62): note: zobacz deklarację „GetTypeInfo<EditorDock::DockLayout,void>”
C:\godot_source\editor/editor_dock.h(77): error C3861: "get_class_info": identyfikatora nie odnaleziono
C:\godot_source\editor/editor_dock.h(77): error C2065: "METADATA": niezadeklarowany identyfikator
scons: *** [bin\obj\editor\debugger\.scu\scu_editor_debugger.gen.windows.editor.dev.x86_64.obj] Error 2
C:\godot_source\editor/editor_dock.h(77): error C2027: wykorzystanie niezdefiniowanego przez użytkownika typu "GetTypeInfo<EditorDock::DockLayout,void>"
C:\godot_source\core/variant/type_info.h(62): note: zobacz deklarację „GetTypeInfo<EditorDock::DockLayout,void>”
C:\godot_source\editor/editor_dock.h(77): error C3861: "get_class_info": identyfikatora nie odnaleziono
C:\godot_source\editor/editor_dock.h(77): error C2065: "METADATA": niezadeklarowany identyfikator
```
I assume no.

kitbdev commented on doc/classes/EditorDock.xml:
```diff
@@ -0,0 +1,106 @@
+		</member>
+	</members>
+	<signals>
+		<signal name="tab_style_changed">
```
I agree, its not very useful for users to emit or respond too.

KoBeWi replied:
Removed.

Mickeon reviewed (commented):
I do still find a bit weird that `available_layouts` exists as a future-proof measure. Docs are fine, now it's a matter to see if the PR is good to go.

KoBeWi commented:
Rebased. I also marked EditorDock as experimental for now.

KoBeWi commented on editor/docks/editor_dock_manager.cpp:
```diff
@@ -253,10 +260,9 @@ Control *EditorDockManager::_get_dock_tab_dragged() {
 			return nullptr;
 		}
 
-		// TODO: Update logic when GH-106503 is merged to cast directly to EditorDock instead of the below check.
```
Ok I see it now. The code is just confusing.

KoBeWi commented on editor/docks/editor_dock_manager.cpp:
```diff
@@ -647,22 +650,23 @@ void EditorDockManager::save_docks_to_config(Ref<ConfigFile> p_layout, const Str
-	for (const KeyValue<Control *, DockInfo> &d : all_docks) {
-		d.key->call(SNAME("_save_layout_to_config"), p_layout, p_section);
+	for (const EditorDock *dock : all_docks) {
+		const String section_name = p_section + "/" + dock->get_effective_layout_key();
+		dock->save_layout_to_config(p_layout, section_name);
```
IMO keeping all layouts in one config file is a bad idea.

kitbdev reviewed (approved):
Looks good.

Repiteo commented:
Thanks!
